### DISCUSSION TITLE
一体化项目如何正确输出带状态码的错误

### DISCUSSION THREAD

**Discussion by xmsz** [midwayjs/midway#1639] [Q&A]
背景
我现在是这样报错
```

export const someApi = async () => {
  const ctx = useContext()
  if(flag) {
    return {message: "ok", data: {a:{b:{c:1}}}
  } else {
    ctx.status = 400;
   return {message: "error ..."}
  }
}

```
如果出错状态码变成400，然后没有输入data

但是这样写，类型提示就会变得很奇怪(data变成了可选)，像这样
```
{
  message: string,
  data?: {a:{b:{c:number}}}
}

```

然后我尝试用Promise.reject，但是不管用

然后我看到普通项目可以通过 throw MidwayError(400, {message:"error"})，但hooks里并没有地方导出MidwayError（我也没自己尝试了）

-------

所以，要如何正确的进行有状态码的报错

类似普通项目

```


someApi().then(
res as {
  message: "", data: {a:{...}}
  }
).catch(
err as {
 response: { 
  status: number, 
  data: {message: string} }
 }
)

```

**Comment by czy88840616** [comment]
报错也是直接tbrow就好，MidwayError是通用的，core里导出的。

**Reply by xmsz** [reply]
好的 我再试一下

**Reply by xmsz** [reply]
我刚刚尝试了一下
    throw new MidwayHttpError('未找到用户', 400);
    throw Error('未找到用户');


- 有改变状态码
- 但没有返回message
![image](https://user-images.githubusercontent.com/5879146/151658005-bebadb72-dc41-4bed-9a99-8840e3c649f3.png)


然后手动加了个一个errorFilter才实现了效果
```
import { Catch } from '@midwayjs/decorator';
import { MidwayHttpError } from '@midwayjs/core';
import { Context } from '@midwayjs/koa';

@Catch()
export class HttpErrorFilter {
  async catch(err: MidwayHttpError, ctx: Context) {
    ctx.logger.error(err);

    ctx.status = err.status ?? 500;

    return {
      code: err.code,
      message: err.message,
    };
  }
}

```

有其他方式快捷的实现吗

**Reply by czy88840616** [reply]
看着没有问题，你看到的错误是兜底的 onerror 返回，由于 error 中的 message 往往含有私密信息， 不可能直接抛出，必须在用户知道的情况下，自己来返回。

**Reply by xmsz** [reply]
> 看着没有问题，你看到的错误是兜底的 onerror 返回，由于 error 中的 message 往往含有私密信息， 不可能直接抛出，必须在用户知道的情况下，自己来返回。

ok
